src/session.ts

```diff
@@ -163,28 +168,30 @@ class NotebookSession {
       if (success.xhr.status !== 204) {
         throw Error('Invalid response');
       }
-      validateSessionId(success.data);
+      validate.validateSessionId(success.data);
     }, (rejected: utils.IAjaxError) => {
-        if (rejected.xhr.status === 410) {
-          throw Error('The kernel was deleted but the session was not');
-        }
-        throw Error(rejected.statusText);
+      if (rejected.xhr.status === 410) {
+        throw Error('The kernel was deleted but the session was not');
+      }
+      throw Error(rejected.statusText);
     });
   }
 
   /**
    * Restart the session by deleting it and then starting it fresh.
    */
-  restart(options?: ISessionOptions): Promise<void> {
-    return this.delete().then(() => this.start()).catch(
-        () => this.start()).then(() => {
-      if (options && options.notebookPath) {
-        this._notebookPath = options.notebookPath;
-      }
-      if (options && options.kernelName) {
-        this._kernel.name = options.kernelName;
-      }
-    })
+  restart(options?: ISessionOptions): Promise<ISessionId> {
+    return this.delete().then(() => {
+      return this.start().then((id: ISessionId) => {
+        if (options && options.notebookPath) {
+          this._notebookPath = options.notebookPath;
+        }
+        if (options && options.kernelName) {
+          this._kernel.name = options.kernelName;
+        }
+        return id;
+      });
+    });
```

Comment:
This block should be tested in general. I'm not sure the original logic was even correct. `this.start()` will never throw, which means the `catch` clause would never execute. I've updated it to what *I think* it should be. 

I don't understand why the options handling is the way it is. But that's probably a questions for @minrk.